db/commitlog/commitlog.cc

```diff
@@ -1418,6 +1418,13 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
 
                 clogger.trace("Pre-writing {} of {} KB to segment {}", (max_size - existing_size)/1024, max_size/1024, filename);
 
+                // re-open without o_dsync for pre-alloc. The reason/rationale
+                // being that we want automatic (meta)data sync from O_DSYNC for when
+                // we do actual CL flushes, but here it would just result in
+                // data being committed before we've reached eof/finished writing.
+                // Again an argument for sendfile-like constructs I guess...
+                co_await f.close();
+                f = co_await open_file_dma(filename, flags & open_flags(~int(open_flags::dsync)), opt);
                 co_await f.allocate(existing_size, max_size - existing_size);
```

Comment:
Why re-open? Isn't this the first open? And we should then re-open with O_DSYNC for actual use.

Also, we should switch back to 128k writes for pre-zero.

Reply:
It is not. As you can clearly see. This is based on the fact that a significant portion of the time, we are opening recycled segments that already are the full size large. So we never have to get to this branch. 

And write size is 128k. Look at line 1436. We write the same buffer N times to form a single 128k writev.

Reply:
But after the re-open, won't we need a re-re-open to restore O_DSYNC?

See 114da51d738b65430cd3ff3bdfce64a46296491d (master) and df2f67626bce617b762c1dfe6f7441eceeb310d7 (reverted).

Reply:
Yes. Line 1454.

Reply:
Ah, ok.

What about the second part? About the 128k stuff that was fixed by @gleb-cloudius and then reverted?

I suppose it's logically another change and can come afterwards.

Reply:
Line 1436. We write up to 128k in each writev. Loop at 1442 builds iovec of same, repeated, buffer.